cranelift/codegen/src/isa/x64/inst/regs.rs

```diff
@@ -33,99 +33,108 @@ fn gpr(enc: u8, index: u8) -> Reg {
 }
 
 pub(crate) fn r12() -> Reg {
-    gpr(ENC_R12, 0)
+    gpr(ENC_R12, 16)
```

Comment:
I just want to check here that these changes to the register numbering don't change the order at which they appear in the final `RealRegUniverse`.  (To be specific: you are not reordering registers within each reg-class, in the universe).  Is that true?  I ask because the ordering of regs within each reg class is relevant -- at least for RA/BT -- to whether caller- or callee-saved registers are used in preference, and that has an effect on the quality of the generated allocation.

Reply:
It's a long story, but... yes, this does put R15 at the end of the allocatable list.

The reason is the following: regalloc has the concept of overall allocatable registers. So if in certain modes (i.e. when the use_pinned_reg flag is true), we want to mark one register as allocatable or not, then it must be at the edge of the "allocatable" list, so at the end of its register class (otherwise its register index changes when the pinned reg is in use).

To keep BT happy with the order of caller or callee-saved registers, I guess we could have a different register index for r15, according to whether it's pinned or not. I think it might cause more churn for other registers (namely, those which will be placed after r15). So maybe this justifies a completely different `RegUniverse` of its own? If that's ok, i'd like to defer this. We may need having different RegUniverse for respecting the win64 calling convention.

Reply:
> If that's ok, i'd like to defer this.

Oh, definitely defer.  Let's get this landed and running first, then we can come back and optimise later.  (Sorry .. I should have made this clearer in the comment)